File: runtimes/go/rabbitmq.go
package runtime

import (
	"context"
	"fmt"
	"log"

	amqp "github.com/rabbitmq/amqp091-go"
)

// RabbitMQAsyncSink provides a fire-and-forget AMQP publisher
type RabbitMQAsyncSink struct {
	conn     *amqp.Connection
	channel  *amqp.Channel
	exchange string
}

func NewRabbitMQAsyncSink(amqpURI, exchange string) (*RabbitMQAsyncSink, error) {
	conn, err := amqp.Dial(amqpURI)
	if err != nil {
		return nil, fmt.Errorf("failed to connect to RabbitMQ: %w", err)
	}

	ch, err := conn.Channel()
	if err != nil {
		conn.Close()
		return nil, fmt.Errorf("failed to open a channel: %w", err)
	}

	err = ch.ExchangeDeclare(
		exchange, // name
		"topic",  // type
		true,     // durable
		false,    // auto-deleted
		false,    // internal
		false,    // no-wait
		nil,      // arguments
	)
	if err != nil {
		ch.Close()
		conn.Close()
		return nil, fmt.Errorf("failed to declare an exchange: %w", err)
	}

	return &RabbitMQAsyncSink{
		conn:     conn,
		channel:  ch,
		exchange: exchange,
	}, nil
}

func (s *RabbitMQAsyncSink) PublishAsync(ctx context.Context, routingKey string, payload []byte) error {
	err := s.channel.PublishWithContext(ctx,
		s.exchange,
		routingKey,
		false, // mandatory
		false, // immediate
		amqp.Publishing{
			ContentType:  "application/octet-stream",
			DeliveryMode: amqp.Persistent,
			Body:         payload,
		})
	if err != nil {
		log.Printf("Failed to publish message: %v", err)
		return err
	}
	return nil
}

func (s *RabbitMQAsyncSink) Close() error {
	if s.channel != nil {
		s.channel.Close()
	}
	if s.conn != nil {
		return s.conn.Close()
	}
	return nil
}

File: runtimes/rust/src/rabbitmq.rs
use lapin::{
    options::*, BasicProperties, Connection, ConnectionProperties, types::FieldTable,
};
use std::sync::Arc;

pub struct RabbitMQAsyncSink {
    connection: Arc<Connection>,
    channel: lapin::Channel,
    exchange: String,
}

impl RabbitMQAsyncSink {
    pub async fn new(amqp_uri: &str, exchange: &str) -> Result<Self, lapin::Error> {
        let conn = Connection::connect(
            amqp_uri,
            ConnectionProperties::default(),
        )
        .await?;

        let channel = conn.create_channel().await?;

        // Declare the exchange if it doesn't exist
        channel
            .exchange_declare(
                exchange.into(),
                lapin::ExchangeKind::Topic,
                ExchangeDeclareOptions {
                    durable: true,
                    ..Default::default()
                },
                FieldTable::default(),
            )
            .await?;

        Ok(Self {
            connection: Arc::new(conn),
            channel,
            exchange: exchange.to_string(),
        })
    }

    pub async fn publish_async(&self, routing_key: &str, payload: &[u8]) -> Result<(), String> {
        let confirm = self.channel
            .basic_publish(
                self.exchange.as_str().into(),
                routing_key.into(),
                BasicPublishOptions::default(),
                payload,
                BasicProperties::default(),
            )
            .await
            .map_err(|e| e.to_string())?;

        let _ = confirm.await.map_err(|e| e.to_string())?;
        Ok(())
    }
}
